### PR TITLE
Feat: packages analysis and benchmark

### PR DESCRIPTION
**Pull request by DomonJi**
image
image
image
image
image

### CODE REVIEW

File: packages/package/src/upload.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+const bundlingPackage = (path: string, outfile: string) => {
+  execSync('npx yarn pack --filename package.tgz')
+  execSync(`mv ${join(path, 'package.tgz')} ${outfile}`)
+}
+
+export const pack = async (path: string, packageJson: PackageJson, options: PackOptions = {}) => {
+  const { name, version } = packageJson
+  const outputDir = `${tmpdir()}/perfsee-package/${`${name}-${version}-${Date.now()}`
```
Comment (EYHN) [inline packages/package/src/upload.ts:76]: https://www.npmjs.com/package/find-cache-dir

File: packages/package/src/cli.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+const readPackageJson = (path: string): PackageJson => {
+  return require(resolve(path, 'package.json')) || {}
+}
+
+const main = async () => {
+  const { _, path = _[0] ?? process.cwd(), platform, project, ...rest } = args
+
+  if (!project) {
+    return console.error(chalk.red('[perfsee] project id required.'))
```
[the remark this answers is not in the document]
Comment (DomonJi) [reply packages/package/src/cli.ts:48]: more config, more confusing

File: packages/package/src/upload.ts (hunk 1 above)
Comment (DomonJi) [reply packages/package/src/upload.ts:89]: I think our usage is not the purpose of this package.

File: packages/package/src/cli.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (EYHN) [reply packages/package/src/cli.ts:48]: feeling tired, if i should typing project over and over again

File: packages/package/src/cli.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { lstatSync } from 'fs'
+import { resolve } from 'path'
+
+import chalk from 'chalk'
+import minimist, { ParsedArgs } from 'minimist'
```
Comment (EYHN) [inline packages/package/src/cli.ts:21]: use `clipanion` , and user can use '--help'

File: packages/package/src/analyzer.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  // @ts-expect-error
+  return console
+}
+
+eventQueue.on('TASK_PACKAGE_STATS', ({ package: pkg }: any) => {
+  version = pkg.version
+  name = pkg.name
+})
+
+eventQueue.on('TASK_PACKAGE_JSON_DETAILS', ({ package: packageJsonDetail }) => {
```
Comment (EYHN) [inline packages/package/src/analyzer.ts:45]: We have forked package-build-stats into our project, can we make those information return from `getPackageStats`

File: packages/package/src/upload.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+      },
+      ['./'],
+    )
+      .pipe(createWriteStream(packPath))
+      .on('finish', () => resolve(packPath))
+      .on('error', (err) => reject(err))
+  })
+}
+
+export const uploadPack = async (
```
Comment (EYHN) [inline packages/package/src/upload.ts:152]: still upload pack even use `--local` flag ?

File: packages/package/src/upload.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
+/*
```
Comment (EYHN) [inline packages/package/src/upload.ts:1]: maybe reorganize this file, it named `upload` but it include `pack`, `analyze`, `pack or not pack`, `upload`.

File: packages/package/package.json [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    "@perfsee/utils": "workspace:*",
+    "@swc-node/register": "^1.6.3",
+    "benchmark": "^2.1.4",
+    "chalk": "^4.1.2",
+    "esbuild": "^0.17.11",
+    "glob": "^9.2.1",
+    "lodash": "^4.17.21",
+    "minimist": "^1.2.8",
+    "node-fetch": "^2.6.2",
+    "platform": "^1.3.6",
```
Comment (EYHN) [inline packages/package/package.json:36]: not used

File: packages/package/tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      "path": "../plugins/utils"
+    },
+    {
+      "path": "../utils"
+    },
+    {
+      "path": "../package-build-stats"
+    },
+    {
+      "path": "../chrome-finder"
```
Comment (EYHN) [inline packages/package/tsconfig.json:27]: delete

File: packages/package/package.json [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  ],
+  "publishConfig": {
+    "registry": "https://registry.npmjs.org/",
+    "access": "public"
+  },
+  "bin": "./bin/cli.js",
+  "dependencies": {
+    "@arrows/composition": "^1.2.2",
+    "@arrows/dispatch": "^1.0.3",
+    "@arrows/multimethod": "^2.1.0",
```
Comment (EYHN) [inline packages/package/package.json:23]: not used

File: packages/package/package.json [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    "@perfsee/chrome-finder": "workspace:*",
+    "@perfsee/package-build-stats": "workspace:*",
+    "@perfsee/plugin-utils": "workspace:*",
+    "@perfsee/utils": "workspace:*",
+    "@swc-node/register": "^1.6.3",
+    "benchmark": "^2.1.4",
+    "chalk": "^4.1.2",
+    "esbuild": "^0.17.11",
+    "glob": "^9.2.1",
+    "lodash": "^4.17.21",
```
Comment (EYHN) [inline packages/package/package.json:33]: not used

File: packages/platform-server/src/db/mysql/project.entity.ts [hunk 11]
```diff
@@ -44,6 +44,7 @@ import type { SourceIssue } from './source-issue.entity'
 import type { Timer } from './timer.entity'
 import type { UserPermission } from './user-permission.entity'
 import type { Webhook } from './webhook.entity'
+import { Package } from './package.entity'
```
Comment (EYHN) [inline packages/platform-server/src/db/mysql/project.entity.ts:47]: type

File: packages/platform-server/src/db/mysql/package.entity.ts [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  /**
+   * This property can be used to query the package list, leftJoin AppVersion table, and store appVersion on this property.
+   */
+  appversion?: AppVersion | null
+}
+
+@Entity()
+export class PackageName extends BaseEntity {
```
Comment (EYHN) [inline packages/platform-server/src/db/mysql/package.entity.ts:201]: why we need this table？

File: packages/platform-server/src/db/mysql/package.entity.ts [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+  @PrimaryGeneratedColumn('increment')
+  id!: number
+
+  @ManyToOne('Project', 'artifacts', { onDelete: 'CASCADE' })
```
Comment (EYHN) [inline packages/platform-server/src/db/mysql/package.entity.ts:205]: artifacts ?

File: packages/package/src/cli.ts (hunk 3 above)
Comment (DomonJi) [reply packages/package/src/cli.ts:21]: clipanion is 10 times larger than minimist

File: packages/package/src/cli.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (DomonJi) [reply packages/package/src/cli.ts:48]: just write in "scripts"

[the remark this answers is not in the document]
Comment (DomonJi) [reply packages/package/src/cli.ts:48]: I really dont want to add something like `perfsee.config`. In `package.json` is fine.

**Review by EYHN — APPROVED** [review]

**Pull request merged**